File: dotnet/src/Pactman.NonprofitCheckPlus/Models/PactmanResult.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Pactman.NonprofitCheckPlus.Models
{
    /// <summary>
    /// Fields shared by every result this SDK returns.
    /// </summary>
    /// <remarks>
    /// Results are shapes this SDK defines, so they are ordinary typed objects — unlike
    /// <see cref="Nonprofit"/>, whose shape the wire dictates.
    /// </remarks>
    public abstract class PactmanResult
    {
        /// <summary>Initializes the fields every result carries.</summary>
        /// <param name="checkCount">Checks consumed so far in the current billing cycle.</param>
        /// <param name="timeTakenMs">Server-side processing time in milliseconds, when reported.</param>
        /// <param name="errors">Item-level failures reported alongside the response.</param>
        /// <param name="requestId">Correlation identifier from the response headers, when present.</param>
        /// <param name="status">HTTP status of the response.</param>
        /// <param name="raw">The unmodified parsed response body.</param>
        protected PactmanResult(
            int? checkCount,
            double? timeTakenMs,
            IReadOnlyList<ApiErrorDetail> errors,
            string? requestId,
            int status,
            ResponseBody raw)
        {
            CheckCount = checkCount;
            TimeTakenMs = timeTakenMs;
            Errors = errors ?? Array.Empty<ApiErrorDetail>();
            RequestId = requestId;
            Status = status;
            Raw = raw ?? ResponseBody.Empty;
        }

        /// <summary>
        /// <c>nonprofit_check_count</c> from the envelope: checks consumed so far in the
        /// current billing cycle, including this request, resetting each cycle.
        /// </summary>
        /// <remarks>
        /// Not the size of this request — take the delta between two responses if you
        /// need that.
        /// </remarks>
        public int? CheckCount { get; }

        /// <summary>Server-side processing time in milliseconds, when reported.</summary>
        public double? TimeTakenMs { get; }

        /// <summary>
        /// Item-level failures reported alongside a successful response. Empty when the
        /// API reported none.
        /// </summary>
        public IReadOnlyList<ApiErrorDetail> Errors { get; }

        /// <summary>
        /// Correlation identifier from the response headers, when the server sent one.
        /// </summary>
        public string? RequestId { get; }

        /// <summary>HTTP status of the response.</summary>
        public int Status { get; }

        /// <summary>
        /// The unmodified parsed response body, including any field not typed above.
        /// </summary>
        /// <remarks>
        /// Normally the parsed envelope; a server that answers 200 with a non-JSON body
        /// leaves the raw text here rather than discarding the evidence.
        /// </remarks>
        public ResponseBody Raw { get; }

        /// <summary>A serializable view of the result, suitable for structured logging.</summary>
        /// <returns>The result's fields, keyed by name.</returns>
        public virtual IReadOnlyDictionary<string, object?> ToDictionary() =>
            new Dictionary<string, object?>(StringComparer.Ordinal)
            {
                ["checkCount"] = CheckCount,
                ["timeTakenMs"] = TimeTakenMs,
                ["errors"] = Errors.Select(detail => detail.ToDictionary()).ToList(),
                ["requestId"] = RequestId,
                ["status"] = Status,
            };
    }
}

File: java/sdk/src/main/java/org/pactman/nonprofitcheckplus/models/PactmanResult.java
package org.pactman.nonprofitcheckplus.models;

import java.util.Collections;
import java.util.List;

/** Fields shared by every result this SDK returns. */
public abstract class PactmanResult {

    private final Integer checkCount;
    private final Double timeTakenMs;
    private final List<ApiErrorDetail> errors;
    private final String requestId;
    private final int status;
    private final ResponseBody raw;

    /**
     * Initializes the shared fields.
     *
     * @param checkCount  checks consumed in the current billing cycle, or {@code null}.
     * @param timeTakenMs server-side processing time, or {@code null}.
     * @param errors      item-level failures the API reported alongside the response.
     * @param requestId   correlation identifier from the response headers, or {@code null}.
     * @param status      the HTTP status of the response.
     * @param raw         the unmodified parsed response body.
     */
    protected PactmanResult(
            Integer checkCount,
            Double timeTakenMs,
            List<ApiErrorDetail> errors,
            String requestId,
            int status,
            ResponseBody raw) {
        this.checkCount = checkCount;
        this.timeTakenMs = timeTakenMs;
        this.errors = errors == null
                ? Collections.<ApiErrorDetail>emptyList()
                : Collections.unmodifiableList(errors);
        this.requestId = requestId;
        this.status = status;
        this.raw = raw;
    }

    /**
     * {@code nonprofit_check_count} from the envelope: checks consumed so far in
     * the current billing cycle, including this request, resetting each cycle.
     *
     * <p>Not the size of this request. Take the delta between two responses if
     * you need that, and read this one as a usage gauge.
     *
     * @return the running total, or {@code null} when the API returned none.
     */
    public Integer getCheckCount() {
        return checkCount;
    }

    /**
     * Server-side processing time in milliseconds, when reported.
     *
     * @return the duration, or {@code null}.
     */
    public Double getTimeTakenMs() {
        return timeTakenMs;
    }

    /**
     * Item-level failures reported alongside a successful response.
     *
     * @return the details. Empty when the API reported none.
     */
    public List<ApiErrorDetail> getErrors() {
        return errors;
    }

    /**
     * Correlation identifier from the response headers, when the server sent one.
     *
     * @return the request id, or {@code null}.
     */
    public String getRequestId() {
        return requestId;
    }

    /**
     * The HTTP status of the response.
     *
     * @return the status code.
     */
    public int getStatus() {
        return status;
    }

    /**
     * The unmodified parsed response body, including any fields not typed above.
     *
     * @return the envelope.
     */
    public ResponseBody getRaw() {
        return raw;
    }
}
